### PR TITLE
release-21.2: backupccl: allow user to specify separate path for incremental backups

### PR DESCRIPTION
**Pull request by msbutler**
Backport 5/5 commits from #72713 , #73357, #73744, #76416, #76274 

/cc @cockroachdb/release

Release justification: customers urgently want to create separate directories for incremental backups and full backups.

### CODE REVIEW

**Comment by dt** [comment]
can you add the redaction and parameter rename commits here so they all backport at once?

File: pkg/ccl/backupccl/backup_planning.go [hunk 1]
```diff
@@ -317,7 +318,7 @@ func getURIsByLocalityKV(to []string, appendPath string) (string, map[string]str
 }
 
 func resolveOptionsForBackupJobDescription(
-	opts tree.BackupOptions, kmsURIs []string,
+	opts tree.BackupOptions, kmsURIs []string, incrementalStorage []string,
```
Comment (adityamaru) [inline pkg/ccl/backupccl/backup_planning.go:321]: nit: rename parameter to incrementalLocation?

File: pkg/ccl/backupccl/backup_planning.go [hunk 2]
```diff
@@ -332,12 +333,16 @@ func resolveOptionsForBackupJobDescription(
...
 	}
 
-	for _, uri := range kmsURIs {
-		redactedURI, err := cloud.RedactKMSURI(uri)
-		if err != nil {
-			return tree.BackupOptions{}, err
-		}
-		newOpts.EncryptionKMSURI = append(newOpts.EncryptionKMSURI, tree.NewDString(redactedURI))
+	var err error
+	// TODO(msbutler): use cloud.RedactKMSURI(uri) here instead?
```
Comment (adityamaru) [inline pkg/ccl/backupccl/backup_planning.go:337]: Does RedactKMSURI differ from the driver in sanitizeURIList? Just want to make sure we are not introducing a regression by switching the redaction method?

File: pkg/ccl/backupccl/create_scheduled_backup.go [hunk 3]
```diff
@@ -448,6 +449,17 @@ func doCreateBackupSchedules(
 	if incRecurrence != nil {
 		chainProtectedTimestampRecords = canChainProtectedTimestampRecords(p, eval)
 		backupNode.AppendToLatest = true
+
+		var incDests []string
+		if eval.incrementalStorage != nil {
```
Comment (adityamaru) [inline pkg/ccl/backupccl/create_scheduled_backup.go:454]: nit: rename to eval.incrementalLocation?

File: pkg/ccl/backupccl/manifest_handling.go [hunk 4]
```diff
@@ -716,10 +722,27 @@ func resolveBackupManifests(
 	} else {
 		// Since incremental layers were *not* explicitly specified, search for any
 		// automatically created incremental layers inside the base layer.
-		prev, err := FindPriorBackups(ctx, baseStores[0], IncludeManifest)
+
+		var incStores []cloud.ExternalStorage
+		if len(incFrom) != 0 {
+			incStores = make([]cloud.ExternalStorage, len(incFrom))
+			for i := range incFrom {
+				store, err := mkStore(ctx, incFrom[i], user)
```
Comment (adityamaru) [inline pkg/ccl/backupccl/manifest_handling.go:730]: nit: ignore it if you want, for consistency to the above if condition can we assign directly to incStores[i], and then defer a close on incStores[i]?

**Review by adityamaru — COMMENTED** [review]
Apologies, maybe some of these comments should have come on the original PRs. I've done a first pass, but I want to do one more. Pushing out the comments I have for now.

If you'd like to open a separate commit on master to address/ignore some of these comments, and then bring that commit over to this backport I'd be good with that too.

**Comment by msbutler** [comment]
no longer getting backported to 21.2

**Pull request closed without merging**